### PR TITLE
feat: conditional formatting

### PR DESCRIPTION
**Pull request by IrakliJani**
Closes: https://github.com/lightdash/lightdash/issues/2081

### Description:

data model for filter configuration (supports follow-up tickets)
```json
{
  // as you can see there is a support for multiple conditional formatting configs
  "conditionalFormattings": [
    {
      // color type will be extended to accept color ranges
      // string | string[]
      "color": "red",
      // there can be multiple rules on a single conditional formatting config
      "rules": [ 
        {
          "values": [50],
          "operator": "lessThan"
        }
      ],
      // target is nullable, thus you can apply conditional formatting to all numeric fields
      "target": { "fieldId": "orders_customer_id" } 
    }
  ]
}
```

https://user-images.githubusercontent.com/962095/211058022-96c7e8ae-c62b-4619-8af9-0124c0d55085.mp4

### CODE REVIEW

**Comment by rephus** [comment]
If we're only going to apply formatting to numbers, I'd exclude other fields from `isNull / isNotNull` rules. 

Screenshot from 2023-01-09 08-58-02

**Comment by rephus** [comment]
Until we can support font color, I think we should autodetect dark background colors and set font to white. 

image
Screenshot from 2023-01-09 08-58-29

**Comment by rephus** [comment]
Conditional formatting is not applied on locked columns 
image

File: packages/frontend/src/components/TableConfigPanel/ConditionalFormatting.tsx [hunk 1]
```diff
@@ -0,0 +1,211 @@
...
+    };
+
+    useEffect(() => {
+        onSetConditionalFormattings(config ? [config] : []);
+    }, [config, onSetConditionalFormattings]);
+
+    // conditional formatting only supports number fields for now
+    const filterConfig = FilterTypeConfig[FilterType.NUMBER];
+
+    // TODO: remove nonsense props
```
Comment (rephus) [inline packages/frontend/src/components/TableConfigPanel/ConditionalFormatting.tsx:111]: TODO

**Comment by TuringLovesDeathMetal** [comment]
Nice! This is great @IrakliJani 😍 

Overall, the functionality is really good! **and** you added selecting which field to apply it to (which is a bonus! I had that as a separate, advanced issue for later on). 

Three small changes I'd suggest:

1. the `+ conditional formatting` button should say `+ add new rule`
2. the `remove conditional formatting` button should say `remove rule` (also, maybe this would look better as just an `x` on the rule? Similar to how we do it for adding/removing filter rules?)
3. Since you have `select fields` already working, I'd say that by default, the conditional formatting applies to _**no**_ columns unless they're selected from the drop-down. (but would love to hear what you think about this one!)

**Review by rephus — COMMENTED** [review]
It looks great, amazing job !

image

I love the fact we can apply the same format to multiple columns. 

There are a few leftover TODOs, not sure if you're planning to finish those in this PR, otherwise can we remove the comments (or just the TODO)

File: packages/frontend/src/components/TableConfigPanel/ConditionalFormatting.tsx (hunk 1 above)
Comment (IrakliJani) [reply packages/frontend/src/components/TableConfigPanel/ConditionalFormatting.tsx:111]: oops, not relevant anymore, forgot to remove the comment

File: packages/frontend/src/components/UserSettings/AppearancePanel/index.tsx [hunk 2]
```diff
@@ -82,24 +45,25 @@ const AppearancePanel: FC = () => {
...
-                        )}
-                        onChange={(colorChange) => {
-                            setColors(
-                                colors.map((c, i) =>
-                                    index === i ? colorChange : c,
-                                ),
-                            );
-                        }}
-                    />
+                    <InputWrapper key={index} label={`Color ${index + 1}`}>
```
[the remark this answers is not in the document]
Comment (IrakliJani) [reply packages/frontend/src/components/UserSettings/AppearancePanel/index.tsx:48]: this is not a refactor, I extracted color input to reuse it in my PR.
what is wrong here is that I should have used another color input which uses react-color which I will change in a separate PR

File: packages/frontend/src/components/TableConfigPanel/ConditionalFormatting.tsx [hunk 3]
```diff
@@ -0,0 +1,211 @@
...
+    const activeFields = useExplorerContext((c) => c.state.activeFields);
+    const visibleActiveNumericFields = useMemo(() => {
+        return explore
+            ? getVisibleFields(explore)
+                  .filter((field) => activeFields.has(fieldId(field)))
+                  .filter((field) => field.type === 'number')
+            : [];
+    }, [explore, activeFields]);
+
+    // TODO: this is intentional to support multiple conditional formattings
```
[the remark this answers is not in the document]
Comment (IrakliJani) [reply packages/frontend/src/components/TableConfigPanel/ConditionalFormatting.tsx:37]: yep, this is a reminder for me for the second PR → issue here https://github.com/lightdash/lightdash/issues/3932

**Comment by IrakliJani** [comment]
@rephus @TuringLovesDeathMetal thanks for the review!!!

stuff to fix:

- [ ] If we're only going to apply formatting to numbers, I'd exclude other fields from isNull / isNotNull rules.
- [ ] Until we can support font color, I think we should autodetect dark background colors and set font to white.
- [ ] Conditional formatting is not applied on locked columns
- [ ] the + conditional formatting button should say + add new rule
- [ ] the remove conditional formatting button should say remove rule (also, maybe this would look better as just an x on the rule? Similar to how we do it for adding/removing filter rules? Or maybe the same red text format we have for deleting charts/dashboards/etc.?)
- [ ] Since you have select fields already working, I'd say that by default, the conditional formatting applies to no columns unless they're selected from the drop-down
- [ ] I would not mind having a separate list for conditional formatting, separate from filters. We can't even support all operators right now.
- [ ] remove todos

**Comment by IrakliJani** [comment]
@rephus made changes and left a note for you in the upper comment

File: packages/frontend/src/utils/colorUtils.ts [hunk 4]
```diff
@@ -0,0 +1,7 @@
+import Color from 'colorjs.io';
+
+export const readableColor = (backgroundColor: string) => {
```
Comment (rephus) [inline packages/frontend/src/utils/colorUtils.ts:3]: nice

File: packages/frontend/src/components/TableConfigPanel/ConditionalFormatting.tsx [hunk 5]
```diff
@@ -84,9 +90,8 @@ const ConditionalFormatting: FC = () => {
 
         setConfig(
             produce(config, (draft) => {
-                // TODO: this is intentional to support multiple rules in the future
                 draft.rules[0] = newRule;
-                // TODO: check if we can fix this problem in number input
+                // FIXME: check if we can fix this problem in number input
```
Comment (rephus) [inline packages/frontend/src/components/TableConfigPanel/ConditionalFormatting.tsx:94]: what's the issue here ? do we need to fix it ?

**Review by rephus — APPROVED** [review]
Code looks good to me, and the feature works very well. 

Perhaps we should not limit the number input to just positive integers. You can still type negative numbers, so it is not a big deal. 

image

:sheep: it !!

File: packages/frontend/src/components/TableConfigPanel/ConditionalFormatting.tsx (hunk 5 above)
Comment (IrakliJani) [reply packages/frontend/src/components/TableConfigPanel/ConditionalFormatting.tsx:94]: I am not sure but sometimes number input returns string. I will deal with that in the following PR or just break out as a separate task

**Comment by IrakliJani** [comment]
Hi @JakeC-MyT absolutely, that makes sense. This pull request is a first stab at a bigger milestone https://github.com/lightdash/lightdash/milestone/63 I'll add an issue in there to support metric columns

**Comment by IrakliJani** [comment]
@rephus ooh, interesting catch. I created a separate issue for that https://github.com/lightdash/lightdash/issues/4153 👍

**Pull request merged**